### PR TITLE
Implement `broadcast_arrays`, `broadcast_to`

### PR DESCRIPTION
**Pull request by neosunhan**
## Description
Implement `broadcast_arrays`, `broadcast_to`

Issue/s resolved: #778

## Changes proposed:
- 

## Type of change
- New feature (non-breaking change which adds functionality)

## Due Diligence
- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

skip ci

### CODE REVIEW

File: heat/core/manipulations.py [hunk 1]
```diff
@@ -89,6 +91,54 @@ def balance(array: DNDarray, copy=False) -> DNDarray:
...
+        DNDarray to broadcast.
+    shape : Tuple[int, ...]
+        Array shape. Must be compatible with x.
+    """
+    if not isinstance(x, DNDarray):
+        raise TypeError("'x' must be a DNDarray, currently {}".format(type(x)))
+
+    broadcasted = torch.broadcast_to(x.larray, shape)
+
+    return factories.array(broadcasted, dtype=x.dtype, device=x.device)
```
Comment (ClaudiaComito) [inline heat/core/manipulations.py:139]:
```suggestion
    # figure out the output split axis via dndarray.__torch_proxy__ and named tensors functionality
    try:
        torch_proxy = x.__torch_proxy__
    except AttributeError:
        raise TypeError("'x' must be a DNDarray, currently {}".format(type(x)))

    split_tags = [None]*x.ndim
    if x.split is not None:
        split_tags[x.split] = "split"
    torch_proxy = torch.tensor(torch_proxy, names=split_tags)
    torch_proxy = torch_proxy.broadcast_to(shape)
    output_split = torch_proxy.names.index("split")

    # exploit binary operations broadcasting 
    broadcasted = factories.zeros(shape, dtype=x.dtype, split=output_split, device = x.device, comm=x.comm)
    broadcasted += x

    return broadcasted
```

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Thanks a lot @neosunhan and apologies again for the Draft PR mix-up.

These two functions don't return the expected values when the input arrays are distributed. I've suggested a change for `broadcast_to` that leverages all the broadcasting checks in `_operations.__bin_op()`. 

You should be able to reimplement `broadcast_arrays` using the new version of `broadcast_to` so that distributed arrays can be broadcasted even if they are distributed along different axes, e.g. this should be possible:
```python
a = ht.arange(100, split=0)
b = ht.arange(200).reshape(2,100)
b.resplit_(axis=1)
# a.split is 0, b.split is 1
ht.broadcast_arrays(a,b)
```

Thanks a lot!

**Review by mrfh92 — CHANGES_REQUESTED** [review]
**Review for PR #1020 "Implement `broadcast_arrays`, `broadcast_to`"**

This PR adds two broadcasting-routines known from NumPy to Heat. 

As far as I can judge, the code is correct and sufficiently commented. However, I would like to suggest to add some more information into the documentation, either by an example what each of the two routines does for a concrete array or by providing reference to the corresponding NumPy-functionality. 

Once these changes have been done, I recommend merging.

File: heat/core/manipulations.py [hunk 2]
```diff
@@ -89,6 +91,81 @@ def balance(array: DNDarray, copy=False) -> DNDarray:
...
+        raise ValueError(e)
+
+    # extract torch tensors
+    t_arrays = list(array.larray for array in arrays)
+
+    broadcasted = torch.broadcast_tensors(*t_arrays)
+
+    out = []
+    for i in range(len(broadcasted)):
+        out.append(factories.array(broadcasted[i], dtype=arrays[i].dtype, device=arrays[i].device))
```
Comment (mrfh92) [inline heat/core/manipulations.py:118]: Maybe I got that wrong, but wasnt there another PR where we wanted to replace the usage of `factories.array` by `DNDarray` for performance reasons?

Comment (ClaudiaComito) [reply heat/core/manipulations.py]: you're right, and the split axis was missing here as well. Fixed.
